src/java.base/share/classes/sun/security/ssl/SSLCipher.java

```diff
@@ -1859,10 +1859,20 @@ public Plaintext decrypt(byte contentType, ByteBuffer bb,
                 }
 
                 if (bb.remaining() <= tagSize) {
-                    throw new BadPaddingException(
-                        "Insufficient buffer remaining for AEAD cipher " +
-                        "fragment (" + bb.remaining() + "). Needs to be " +
-                        "more than tag size (" + tagSize + ")");
+                    // Check for unexpected plaintext alert.
+                    if (contentType == ContentType.ALERT.id
+                        && bb.remaining() == 2) {
+                        throw new GeneralSecurityException(String.format(
```

Comment:
- It would be a flaw of protocol implementation of the other party, so if we are the server the flaw would be on the client's side which makes sense in this scenario.
- Yes, I've decided to use `GeneralSecurityException` for all the reasons you have listed.